### ISSUE TITLE
Question about symbol missing in mcount_arch_parent_location

### ISSUE THREAD

**Issue by nadongguri** [namhyung/uftrace#1629]
Dear @namhyung, and @honggyukim, 

I got an issue when I try to trace our program by using uftrace. :)
I'd like to solve this by debugging more, but I'm a newbie of uftrace.
Is it possible to get a little tip on where to start? 

**Situation**
The program I'd like to trace is a daemon implemented based on gobject running in Tizen TV.
That source code is quite long so I used to trace it long ago. (it's super useful!)
The issue is occured with the latest version of the program. 
(I tried to trace it with the latest of uftrace and the old one that I compiled long ago, both are failed.)

**Error message**
The error message is 
```
uftrace: uftrace/arch/arm/mcount-support.c:248:mcount_arch_parent_location
 ERROR: cannot find symbol for b539c939
```
I found that it seems to occur in the process of calling the g_module_open function.

**Observation**
Actually, ```-flto``` option is enabled by default in our build env but when I try to build uftrace with that option, there are ```undefined reference```  errors to functions written in assembly. 
So as an exception, I built and used uftrace except for this option.

**System Info**
Platform : Tizen-TV
Architecture : ARM
Linux : 5.4.77

With the best regards,
Dongjin.

**Comment by honggyukim** [comment]
Hi @nadongguri, it's great to see you here :)

The error is shown at https://github.com/namhyung/uftrace/blob/v0.13.1/arch/arm/mcount-support.c#L243-L245.

So one thing we can do now is to add print statements inside `find_symtabs` where it returns `NULL` and see where exactly causes the problem.
https://github.com/namhyung/uftrace/blob/v0.13.1/utils/symbol.c#L1462-L1505

It would also be nice if you could share the disassembled binary instructions for the `g_module_open` function. You don't have to share the entire function but we only need a few instructions of the function prologue before and after calling `__gnu_mcount_nc` or `mcount`.

There might be some issues if the binary uses ARM and Thumb modes together. We need more testing but couldn't test it enough these days.

**Comment by namhyung** [comment]
I suspect the address itself is invalid.  Could you please verify that `b539c939` is a valid address?  If you have a uftrace.data and the uftrace source tree, you can run `misc/symbols` command (you can just copy the binary).

```
$ misc/symbols 0xb539c939
```

And I also recommend to use `--estimate-return` option, but then realized it should have this too

```diff
diff --git a/libmcount/mcount.c b/libmcount/mcount.c
index 64693fd1..e45d585a 100644
--- a/libmcount/mcount.c
+++ b/libmcount/mcount.c
@@ -1330,7 +1330,8 @@ static int __mcount_entry(unsigned long *parent_loc, unsigned long child, struct
        }
 
        /* fixup the parent_loc in an arch-dependant way (if needed) */
-       parent_loc = mcount_arch_parent_location(&mcount_sym_info, parent_loc, child);
+       if (!mcount_estimate_return)
+               parent_loc = mcount_arch_parent_location(&mcount_sym_info, parent_loc, child);
 
        if (mcount_estimate_return)
                mcount_rstack_inject_return(mtdp, parent_loc, child);
```

**Comment by nadongguri** [comment]
Thank you for your kind and helpful reply.
I would like to inform you that it may be difficult to share the necessary information because the program is a closed source.

**Report**
@namhyung ,
I'm on the same opinion. Here's the result.
```
root@localhost:~# uftrace record /bin/npumgr_server_
uftrace: /home/abuild/rpmbuild/BUILD/uftrace-0.13.1/arch/arm/mcount-support.c:245:mcount_arch_parent_location
 ERROR: cannot find symbol for b4f65e03
root@localhost:~# symbols 0xb4f65e03
b4f65e03:
```
When I checked sid-xxx.map in uftrace.data, I think that address is invalid.
```
43890000-438be000 r-xp 00000000 00:00 0
46c50000-46c66000 r-xp 00000000 00:00 0
// No libs cover b4f65e03 address
b60c3000-b60d7000 r-xp 00000000 00:00 0
b60d7000-b610c000 r-xp 00000000 00:00 0
b612d000-b6141000 r-xp 00000000 00:00 0
be13d000-be15e000 rw-p 00000000 00:00 0
```

Also, applying that patch and run uftrace with --estimate-return option works. :)
But carefully speaking, I don't think the tracing works well. (The functions to be recorded are not visible when replaying. )
I will try to debug while analyzing the code by applying the method proposed by @honggyukim . :)


**g_module_open**
This function is in libgmodule.so. I didn't rebuild that library by adding ```-pg -finstrument-functions``` options when I raised an issue. 
So, I put flags, and built it, and tested it. The result is the same.

Here is disassembled code of g_module_open function.
```
000026d0 <g_module_open>:
    26d0:   b5f8        push    {r3, r4, r5, r6, r7, lr}
    26d2:   b500        push    {lr}
    26d4:   f7fe ebcc   blx e70 <__gnu_mcount_nc@plt>
    26d8:   4b0f        ldr r3, [pc, #60]   ; (2718 <g_module_open+0x48>)
    26da:   4605        mov r5, r0
    26dc:   4a0f        ldr r2, [pc, #60]   ; (271c <g_module_open+0x4c>)
    26de:   460e        mov r6, r1
    26e0:   447b        add r3, pc
    26e2:   4671        mov r1, lr
    26e4:   4674        mov r4, lr
    26e6:   589f        ldr r7, [r3, r2]
    26e8:   4638        mov r0, r7
    26ea:   f7fe eb00   blx cec <__cyg_profile_func_enter@plt>
    ///
    2700:   f7fe eb08   blx d14 <__cyg_profile_func_exit@plt+0x4>
    2704:   4620        mov r0, r4
    2706:   bdf8        pop {r3, r4, r5, r6, r7, pc}
    2708:   4621        mov r1, r4
    270a:   4604        mov r4, r0
    270c:   4638        mov r0, r7
    270e:   f7fe eb02   blx d14 <__cyg_profile_func_exit@plt+0x4>
    2712:   4620        mov r0, r4
    2714:   f7fe ebbe   blx e94 <_Unwind_Resume@plt>
    2718:   0001091c    .word   0x0001091c
    271c:   000000d8    .word   0x000000d8
```


**misc**
current system features are 
```
...         prefix: /usr/local
...         libelf: [ OFF ] - more flexible ELF data handling
...          libdw: [ OFF ] - DWARF debug info support
...      libpython: [ OFF ] - python scripting support
...      libluajit: [ OFF ] - luajit scripting support
...    libncursesw: [ OFF ] - TUI support
...   cxa_demangle: [ on  ] - full demangler support with libstdc++
...     perf_event: [ on  ] - perf (PMU) event support
...       schedule: [ OFF ] - scheduler event support
...       capstone: [ OFF ] - full dynamic tracing support
...      libunwind: [ OFF ] - stacktrace support (optional for debugging)
```

**Comment by namhyung** [comment]
Hmm.. the function has both `__gnu_mcount_nc` and `__cyg_profile_func_{enter,exit}` which is weird.

**Comment by honggyukim** [comment]
> Hmm.. the function has both __gnu_mcount_nc and __cyg_profile_func_{enter,exit} which is weird.

@nadongguri Please check whether your binary is compiled with `-pg` and `-finstrument-functions` together. We shouldn't use them together.

> ```
> uftrace: uftrace/arch/arm/mcount-support.c:248:mcount_arch_parent_location
>  ERROR: cannot find symbol for b539c939
> ```

The problem is shown when handling `__gnu_mcount_nc`, which is inserted by `-pg` option. You won't see the same error if you trace your binary with `-finstrument-functions` only. Can you test it again?

**Comment by nadongguri** [comment]
> The problem is shown when handling __gnu_mcount_nc, which is inserted by -pg option. You won't see the same error if you trace your binary with -finstrument-functions only. Can you test it again?

Yes. It's the problem both options are applied! I'm sorry for the trouble.
In summary, that reported S/W package (binary and library) only works with ```-finstrument-functions``` flag at our env!
I will share the spec file required for gbs build so that others can use uftrace within the company.

BTW, is it helpful to analyze more why ```-pg``` flag causes errors?
I did a little experiment with that.

```
$uftrace record /bin/program
uftrace: /home/abuild/rpmbuild/BUILD/uftrace-0.13.1/arch/arm/mcount-support.c:245:mcount_arch_parent_location
ERROR: cannot find symbol for b60e9643
```

That ```program``` binary loads ```library.so``` at runtime using ```g_module_open()``` and calls ```init()``` function in that library using ```g_module_symbol()```.

In the first report, there was no log after the ```g_module_open()``` function call, so I thought the function was a problem, but it seems that the problem occurs when calling the ```init()``` function.

```
uftrace.data/task.txt
DLOP timestamp=16265.384294766 tid=12642 sid=89c5a639237b99a0 base=b60e7000 libname="/usr/lib/library.so"

error symbol address is 0xb60e9643.
The offset is 0x2643

uftrace.data/library.so.sym

00000000000024c4 P __cxa_finalize
00000000000024d0 P __gnu_mcount_nc
00000000000024dc P function_a
00000000000024e8 ? __dynsym_end
000000000000261d T function_b
0000000000002629 T function_c
000000000000263d T init                         // offset gap is 0x6
0000000000002649 T function_d
```

There seems to be a cause of offset going wrong somewhere.

**Comment by honggyukim** [comment]
> Yes. It's the problem both options are applied! I'm sorry for the trouble.

Not at all. I'm glad that it works with `-finstrument-functions`.

> I will share the spec file required for gbs build so that others can use uftrace within the company.

That might be useful. Thanks.

> BTW, is it helpful to analyze more why -pg flag causes errors?

Hmm.. I don't have ARM 32-bit environment so it's difficult for me to debug only with the log. Is it possible to write a simple and short program that reproduces the same problem?

**Comment by nadongguri** [comment]
> Hmm.. I don't have ARM 32-bit environment so it's difficult for me to debug only with the log. Is it possible to write a simple and >short program that reproduces the same problem?

Yes. I can reproduce this error when the program loads library with ```-pg``` using ```g_module_open()``` and calls the function of that library to ```g_module_symbol()```.

I'll check more and get back to you.

**Comment by namhyung** [comment]
IIRC ARM32 thumb2 mode didn't gurarantee the frame pointer to be setup correctly.  Sometimes it has an invalid value and results in crash due to the memory access to the invalid address.  That's why `--estimate-return` didn't crash.  Maybe specifying `-fno-omit-frame-pointer` could help.
